app/controllers/keywords_controller.rb

```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+require 'csv'
+
+class KeywordsController < ApplicationController
+  before_action :ensure_authentication
+  before_action :fetch_file, only: :create
+  before_action :validate_file_type, only: :create
+  before_action :validate_csv, only: :create
+
+  def new
+    render locals: {
+      keyword: Keyword.new
+    }
+  end
+
+  def create
+    Keyword.transaction do
+      CSV.foreach(@file.path, headers: true) do |row|
+        current_user.keywords.create(keyword: row[0])
+      end
+    end
+
+    redirect_to keywords_path, notice: t('keyword.upload_csv_successfully')
+  end
+
+  private
+
+  def fetch_file
+    @file = params[:keyword][:file]
+  end
+
+  def validate_file_type
+    file_type = @file.content_type
+
+    redirect_to keywords_path, alert: t('keyword.file_must_be_csv') unless file_type == 'text/csv'
+  end
```

Comment:
How about creating an FileUploadForm class or sth like that to handle this validation?

Reply:
Could you please explain more about this?
What is the benefit to use class and how I can use class for validation in this case.

Thanks

Reply:
I'm sorry, my comment should be clearer. I meant, the logic for handling the uploaded file, validate it & import the content is quite complex to process in the controller, we should abstracting it to a form. We have [that convention](https://compass.nimblehq.co/development/code-conventions/ruby-on-rails/) defined on Compass.
note that, in this case, we can define a [custom validator](https://compass.nimblehq.co/development/code-conventions/ruby-on-rails/#validators) as well

so how about we extract the CSV import code as a form, e.g.
```ruby
class CsvImportForm
  include ActiveModel::Model

  attr_accessor :user
  validates_with CsvFileValidator

  def initialize(user)
    @user = user
  end

  def save
    return false unless valid?
     # import csv
  end
end

class CsvFileValidator  < ActiveModel::Validator
  def validate(record)
        # given `file` is the csv file field name
        return if record.file.class == ActionDispatch::Http::UploadedFile
        return if record.file.content_type == 'text/scv'

        record.errors.add(:base, 'invalid file')
      end
end

class KeywordsController
  def create
    form = CsvImportForm.new(create_params)

    if form.save
       render ...
    else
       render ...
  end

  private

  def create_params
     params.permit(:file)
end
```

Reply:
Separate form out of controller in fb6fc99bf0a1c846597ee3d68c156b4efbd147c5
Still stuck at validator.

Reply:
Validator fixed in f6b944240f40823b8068275f2fc8ca05590f4ce5